### PR TITLE
vim-patch:8.1.{2005,2010}: the regexp.c file is too big

### PR DESCRIPTION
**Pull request by kylo252**
#### vim-patch:8.1.2005: the regexp.c file is too big

Problem:    The regexp.c file is too big.
Solution:   Move the backtracking engine to a separate file. (Yegappan
            Lakshmanan, closes vim/vim#4905)
https://github.com/vim/vim/commit/6d7d7cf750bca5d641e464f6a3af5ee5b99a5ac8

#### vim-patch:8.1.2010: new file uses old style comments

Problem:    New file uses old style comments.
Solution:   Change to new style comments. (Yegappan Lakshmanan, closes vim/vim#4910)
https://github.com/vim/vim/commit/9490b9a61cf1f1f3fa9758663a33124ea9f71c87

### CODE REVIEW

**Comment by kylo252** [comment]
This PR is basically splitting up a significant chunk of `regexp.c` into `regexp_bt.c`. I've avoided doing any formatting, and also tried to maintain the current order of both files from upstream to make it easier for reviews and future patches.

**Comment by zeertzjq** [comment]
Hmm, there are merge conflicts are #16526 was merged.

File: src/nvim/regexp.c [hunk 1]
```diff
@@ -3,45 +3,6 @@
 
 // uncrustify:off
 
-/*
- * Handling of regular expressions: vim_regcomp(), vim_regexec(), vim_regsub()
```
Comment (zeertzjq) [inline src/nvim/regexp.c]: The original patch didn't delete this line.

Comment (zeertzjq) [reply src/nvim/regexp.c]: Hmm why does GitHub mark this diff as outdated?

File: src/nvim/regexp.c [hunk 2]
```diff
@@ -755,15 +351,29 @@ static int nextchr;             /* used for ungetchr() */
...
+     int	curchr;
+     int	prevchr;
+     int	prevprevchr;
+     int	nextchr;
+     int	at_start;
+     int	prev_at_start;
+     int	regnpar;
+} parse_state_T;
+
+static void initchr(char_u *str);
```
Comment (zeertzjq) [inline src/nvim/regexp.c]: This line doesn't seem necessary.

File: src/nvim/regexp.c [hunk 3]
```diff
@@ -1219,4400 +523,723 @@ char_u *skip_regexp(char_u *startp, int dirc, int magic, char_u **newp)
...
-    else
-      EMSG_RET_NULL(_(e_trailing));             /* "Can't happen". */
-    /* NOTREACHED */
-  }
-  // Here we set the flag allowing back references to this set of
-  // parentheses.
-  if (paren == REG_PAREN) {
-    had_endbrace[parno] = true;  // have seen the close paren
+/// Used in a place where no * or \+ can follow.
+static bool re_mult_next(char *what)
```
Comment (zeertzjq) [inline src/nvim/regexp.c]: Is there is reason why this function is put here instead of between `match_with_backref()` and `decomp_T` like upstream?

Comment (kylo252) [reply src/nvim/regexp.c]: I don't remember if I had any specific reason. I may have shifted it during a rebase or something. I'll move it tho.

File: src/nvim/regexp.c [hunk 4]
```diff
@@ -1219,4400 +523,723 @@ char_u *skip_regexp(char_u *startp, int dirc, int magic, char_u **newp)
...
+/// Wrapper around strchr which accounts for case-insensitive searches and
+/// non-ASCII characters.
+///
+/// This function is used a lot for simple searches, keep it fast!
+///
+/// @param  s  string to search
+/// @param  c  character to find in @a s
+///
+/// @return  NULL if no match, otherwise pointer to the position in @a s
+static inline char_u *cstrchr(const char_u *const s, const int c)
```
[the remark this answers is not in the document]
Comment (kylo252) [reply src/nvim/regexp.c]: Fixed!

File: src/nvim/regexp.c [hunk 5]
```diff
@@ -1219,4400 +526,670 @@ char_u *skip_regexp(char_u *startp, int dirc, int magic, char_u **newp)
...
-    *--dst = *--src;
-
-  place = opnd;                 /* Op node, where operand used to be. */
-  *place++ = op;
-  *place++ = NUL;
-  *place = NUL;
-}
+/* Save the sub-expressions before attempting a match. */
+#define save_se(savep, posp, pp) \
+  REG_MULTI ? save_se_multi((savep), (posp)) : save_se_one((savep), (pp))
```
Comment (zeertzjq) [inline src/nvim/regexp.c]: These lines should be moved to `regexp_bt.c`

File: src/nvim/regexp.c [hunk 6]
```diff
@@ -1219,4400 +526,670 @@ char_u *skip_regexp(char_u *startp, int dirc, int magic, char_u **newp)
...
-{
-  char_u      *src;
-  char_u      *dst;
-  char_u      *place;
+/* After a failed match restore the sub-expressions. */
+#define restore_se(savep, posp, pp) { \
+    if (REG_MULTI) \
+      *(posp) = (savep)->se_u.pos; \
+    else \
+      *(pp) = (savep)->se_u.ptr; }
```
Comment (zeertzjq) [inline src/nvim/regexp.c]: and these

**Review by zeertzjq — CHANGES_REQUESTED** [review]

File: src/nvim/regexp.c [hunk 7]
```diff
@@ -1219,4400 +526,670 @@ char_u *skip_regexp(char_u *startp, int dirc, int magic, char_u **newp)
...
-          }
-          break;
-
-        case END:
-          status = RA_MATCH;    /* Success! */
-          break;
-
-        default:
-          iemsg(_(e_re_corr));
 #ifdef REGEXP_DEBUG
```
[the remark this answers is not in the document]
Comment (kylo252) [reply src/nvim/regexp.c]: I don't see this one anymore, was it this one? https://github.com//kylo252/neovim/blob/b43ade5d71d14dce27522bcb09c0aa4842a8ff21/src/nvim/regexp_bt.c#L3836-L3837

[the remark this answers is not in the document]
Comment (kylo252) [reply src/nvim/regexp.c]: I just did a rebase on master as well for faster builds.

File: src/nvim/regexp.c [hunk 8]
```diff
@@ -1219,4400 +526,654 @@ char_u *skip_regexp(char_u *startp, int dirc, int magic, char_u **newp)
...
-
-  rex.input = scan;
-
-  return (int)count;
+  return true;
 }
 
+#define ADVANCE_REGINPUT() MB_PTR_ADV(rex.input)
+
 /*
```
Comment (zeertzjq) [inline src/nvim/regexp.c]: These lines should be moved to `regexp_bt.c`

File: src/nvim/regexp.c [hunk 9]
```diff
@@ -1219,4400 +526,654 @@ char_u *skip_regexp(char_u *startp, int dirc, int magic, char_u **newp)
...
- * regnext - dig the "next" pointer out of a node
- * Returns NULL when calculating size, when there is no next item and when
- * there is an error.
+ * The arguments from BRACE_LIMITS are stored here.  They are actually local
+ * to regmatch(), but they are here to reduce the amount of stack space used
+ * (it can be called recursively many times).
  */
```
Comment (zeertzjq) [inline src/nvim/regexp.c]: and these

File: src/nvim/regexp.c [hunk 10]
```diff
@@ -1219,4400 +526,654 @@ char_u *skip_regexp(char_u *startp, int dirc, int magic, char_u **newp)
...
-  int offset;
-
-  if (p == JUST_CALC_SIZE || reg_toolong)
-    return NULL;
-
-  offset = NEXT(p);
-  if (offset == 0)
-    return NULL;
+static long bl_minval;
+static long bl_maxval;
```
Comment (zeertzjq) [inline src/nvim/regexp.c]: and these

**Review by zeertzjq — CHANGES_REQUESTED** [review]

**Comment by zeertzjq** [comment]
Many comments where //changes from `/* */` style to `//` style in the original patch.

**Comment by kylo252** [comment]
> Many comments were changed from `/* */` style to `//` style in the original patch.

I'm fairly confident I didn't change them manually (https://github.com/neovim/neovim/pull/16336#issuecomment-971774966), so some of them were already changed before I even moved them. That said, I can do a formatting patch separately, or apply over this one if you'd like.

**Comment by zeertzjq** [comment]
I said the original patch.

**Comment by zeertzjq** [comment]
Oh it is actually not that many.

**Comment by kylo252** [comment]
> Oh it is actually not that many.

Here's what I mean (from master): https://github.com//neovim/neovim/blob/046950f6309070d98ab5a3deeff9d00e079ccdd4/src/nvim/regexp.c#L2536-L2552

**Comment by zeertzjq** [comment]
I said some comments were changed in the original patch in Vim.

File: src/nvim/regexp_bt.c [hunk 11]
```diff
@@ -0,0 +1,5002 @@
...
+ * Uses curbuf for line count and 'iskeyword'.
+ * If "line_lbr" is true, consider a "\n" in "line" to be a line break.
+ *
+ * Returns 0 for failure, number of lines contained in the match otherwise.
+ */
+static int 
+bt_regexec_nl (
+    regmatch_T *rmp,
+    char_u *line,      /* string to match against */
+    colnr_T col,       /* column to start looking for match */
```
Comment (zeertzjq) [inline src/nvim/regexp_bt.c]: These two for example

**Comment by clason** [comment]
> but it felt weird to revert those comment style, since the patch was basically extracting the functions and not really doing anything else.

If you touch it, it's your job to placate lint. Them's the rules. (So, yes, any code you move needs to be adapted to _Neovim's_ code style.)

**Comment by kylo252** [comment]
> If you touch it, it's your job to placate lint. Them's the rules. (So, yes, any code you move needs to be adapted to _Neovim's_ code style.)

Yeah no argument there, I just wanted some confirmation before partaking in that effort. It's also easier now that I have at least one review okaying other non-stylistic changes.

**Comment by zeertzjq** [comment]
I think clason wanted you to passed lint CI?

**Comment by clason** [comment]
Indeed I did.

**Comment by clason** [comment]
> That my not feasible for a change this big.

Of course it's feasible. The question is just whether you're prepared to do the work.

**Comment by kylo252** [comment]
> Of course it's feasible. The question is just whether you're prepared to do the work.

It might be feasible, but I don't have the bandwidth to sink a lot of hours into fixing stylistic violations, also see: https://github.com/neovim/neovim/pull/14895#issuecomment-867986209

**Review by zeertzjq — APPROVED** [review]
I made some other changes so that patch 8.1.2010 can also be marked as ported and squashed.

**Comment by kylo252** [comment]
> I made some other changes so that patch 8.1.2010 can also be marked as ported and squashed.

@zeertzjq, how should we coordinate with https://github.com/neovim/neovim/pull/16730? maybe cherry-pick that commit on top?

**Comment by zeertzjq** [comment]
Maybe.

**Comment by kylo252** [comment]
> Maybe.

Actually the only relevant one to `regexp.c` from there is 8.1.0748. I should also mention that 8.1.1414 and 8.1.1827 are (probably) N/A.

cc: @bfredl

**Comment by zeertzjq** [comment]
That PR contains only one commit.

**Comment by zeertzjq** [comment]
I don't think this should pass lint CI. The clint.py error list is generated around 05:00 UTC every day, so if this is merged some minutes before 05:00 UTC, it should not cause much trouble for other PRs. Considering <https://github.com/neovim/neovim/wiki/Refactoring#frozen-legacy-modules>, @clason what do you think?

**Comment by jamessan** [comment]
> I'm not sure if that link refers to lint fixes while porting patches; I read it more as avoiding gratuitous refactors on its own.

Exactly.

**Pull request merged**